File: scripts/fetch_suggestions.py
#!/usr/bin/env python3
"""
Script to fetch all wellness suggestions from the Emotional Intelligence API.
This demonstrates how to programmatically access all suggestions.
"""

import requests
import json
from typing import Dict, List

def fetch_all_suggestions(api_url: str = "http://localhost:5001") -> Dict:
    """
    Fetch all wellness suggestions from the API.
    
    Args:
        api_url: The base URL of the API (default: http://localhost:5001)
    
    Returns:
        Dictionary containing all suggestion categories and their suggestions
    """
    try:
        response = requests.get(f"{api_url}/suggestions")
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"Error fetching suggestions: {e}")
        return {}

def print_suggestions(suggestions: Dict):
    """
    Print all suggestions in a formatted way.
    
    Args:
        suggestions: Dictionary containing suggestion categories
    """
    if not suggestions or 'categories' not in suggestions:
        print("No suggestions found or invalid response format.")
        return
    
    print("🌟 Wellness Suggestions 🌟")
    print("=" * 50)
    
    for category in suggestions['categories']:
        print(f"\n{category['title']}")
        print("-" * len(category['title']))
        
        for i, suggestion in enumerate(category['suggestions'], 1):
            print(f"{i}. {suggestion}")
    
    print(f"\n📊 Total Categories: {len(suggestions['categories'])}")
    total_suggestions = sum(len(cat['suggestions']) for cat in suggestions['categories'])
    print(f"📝 Total Suggestions: {total_suggestions}")

def get_suggestions_by_category(suggestions: Dict, category_title: str) -> List[str]:
    """
    Get suggestions for a specific category.
    
    Args:
        suggestions: Dictionary containing suggestion categories
        category_title: Title of the category to filter by
    
    Returns:
        List of suggestions for the specified category
    """
    for category in suggestions.get('categories', []):
        if category['title'] == category_title:
            return category['suggestions']
    return []

def get_random_suggestion(suggestions: Dict) -> str:
    """
    Get a random suggestion from any category.
    
    Args:
        suggestions: Dictionary containing suggestion categories
    
    Returns:
        A random suggestion string
    """
    import random
    
    all_suggestions = []
    for category in suggestions.get('categories', []):
        all_suggestions.extend(category['suggestions'])
    
    if all_suggestions:
        return random.choice(all_suggestions)
    return "No suggestions available."

def main():
    """Main function to demonstrate fetching and using suggestions."""
    print("🔍 Fetching wellness suggestions from the API...")
    
    # Fetch all suggestions
    suggestions = fetch_all_suggestions()
    
    if not suggestions:
        print("❌ Failed to fetch suggestions. Make sure the API is running.")
        return
    
    # Print all suggestions
    print_suggestions(suggestions)
    
    # Example: Get suggestions for a specific category
    print("\n" + "=" * 50)
    print("🎯 Example: Stress Management Suggestions")
    print("=" * 50)
    stress_suggestions = get_suggestions_by_category(suggestions, "💪 Stress Management")
    for i, suggestion in enumerate(stress_suggestions, 1):
        print(f"{i}. {suggestion}")
    
    # Example: Get a random suggestion
    print("\n" + "=" * 50)
    print("🎲 Random Suggestion of the Day")
    print("=" * 50)
    random_suggestion = get_random_suggestion(suggestions)
    print(f"💡 {random_suggestion}")
    
    # Example: Save suggestions to a JSON file
    print("\n" + "=" * 50)
    print("💾 Saving suggestions to file...")
    try:
        with open('wellness_suggestions.json', 'w') as f:
            json.dump(suggestions, f, indent=2)
        print("✅ Suggestions saved to 'wellness_suggestions.json'")
    except Exception as e:
        print(f"❌ Error saving suggestions: {e}")

if __name__ == "__main__":
    main() 

File: scripts/fetch_suggestions.js
#!/usr/bin/env node
/**
 * JavaScript script to fetch all wellness suggestions from the Emotional Intelligence API.
 * This demonstrates how to programmatically access all suggestions using Node.js.
 */

const https = require('https');
const http = require('http');
const fs = require('fs');

/**
 * Fetch all wellness suggestions from the API.
 * @param {string} apiUrl - The base URL of the API (default: http://localhost:5001)
 * @returns {Promise<Object>} Dictionary containing all suggestion categories and their suggestions
 */
async function fetchAllSuggestions(apiUrl = "http://localhost:5001") {
    return new Promise((resolve, reject) => {
        const url = new URL(`${apiUrl}/suggestions`);
        const client = url.protocol === 'https:' ? https : http;
        
        const req = client.get(url, (res) => {
            let data = '';
            
            res.on('data', (chunk) => {
                data += chunk;
            });
            
            res.on('end', () => {
                try {
                    const suggestions = JSON.parse(data);
                    resolve(suggestions);
                } catch (error) {
                    reject(new Error(`Failed to parse JSON: ${error.message}`));
                }
            });
        });
        
        req.on('error', (error) => {
            reject(new Error(`Request failed: ${error.message}`));
        });
        
        req.setTimeout(5000, () => {
            req.destroy();
            reject(new Error('Request timeout'));
        });
    });
}

/**
 * Print all suggestions in a formatted way.
 * @param {Object} suggestions - Dictionary containing suggestion categories
 */
function printSuggestions(suggestions) {
    if (!suggestions || !suggestions.categories) {
        console.log("No suggestions found or invalid response format.");
        return;
    }
    
    console.log("🌟 Wellness Suggestions 🌟");
    console.log("=".repeat(50));
    
    suggestions.categories.forEach(category => {
        console.log(`\n${category.title}`);
        console.log("-".repeat(category.title.length));
        
        category.suggestions.forEach((suggestion, index) => {
            console.log(`${index + 1}. ${suggestion}`);
        });
    });
    
    console.log(`\n📊 Total Categories: ${suggestions.categories.length}`);
    const totalSuggestions = suggestions.categories.reduce((sum, cat) => sum + cat.suggestions.length, 0);
    console.log(`📝 Total Suggestions: ${totalSuggestions}`);
}

/**
 * Get suggestions for a specific category.
 * @param {Object} suggestions - Dictionary containing suggestion categories
 * @param {string} categoryTitle - Title of the category to filter by
 * @returns {Array<string>} List of suggestions for the specified category
 */
function getSuggestionsByCategory(suggestions, categoryTitle) {
    const category = suggestions.categories.find(cat => cat.title === categoryTitle);
    return category ? category.suggestions : [];
}

/**
 * Get a random suggestion from any category.
 * @param {Object} suggestions - Dictionary containing suggestion categories
 * @returns {string} A random suggestion string
 */
function getRandomSuggestion(suggestions) {
    const allSuggestions = suggestions.categories.flatMap(cat => cat.suggestions);
    if (allSuggestions.length > 0) {
        const randomIndex = Math.floor(Math.random() * allSuggestions.length);
        return allSuggestions[randomIndex];
    }
    return "No suggestions available.";
}

/**
 * Get suggestions by emotion (mood-based filtering).
 * @param {Object} suggestions - Dictionary containing suggestion categories
 * @param {string} emotion - The emotion to filter suggestions for
 * @returns {Array<string>} Filtered suggestions based on emotion
 */
function getSuggestionsByEmotion(suggestions, emotion) {
    const emotionMap = {
        'sadness': ['🤝 Social Connection', '💡 Creative Expression', '🧘‍♀️ Mindfulness & Relaxation'],
        'anger': ['💪 Stress Management', '🧘‍♀️ Mindfulness & Relaxation', '🏃‍♀️ Physical Wellness'],
        'fear': ['🧘‍♀️ Mindfulness & Relaxation', '🤝 Social Connection', '💪 Stress Management'],
        'joy': ['🎯 Personal Growth', '💡 Creative Expression', '🏃‍♀️ Physical Wellness'],
        'love': ['🤝 Social Connection', '💡 Creative Expression', '🎯 Personal Growth'],
        'surprise': ['💡 Creative Expression', '🎯 Personal Growth', '🏃‍♀️ Physical Wellness'],
        'disgust': ['🧘‍♀️ Mindfulness & Relaxation', '💪 Stress Management', '🌅 Morning Wellness'],
        'neutral': ['🎯 Personal Growth', '🏃‍♀️ Physical Wellness', '💡 Creative Expression']
    };
    
    const relevantCategories = emotionMap[emotion] || [];
    const relevantSuggestions = [];
    
    relevantCategories.forEach(categoryTitle => {
        const categorySuggestions = getSuggestionsByCategory(suggestions, categoryTitle);
        relevantSuggestions.push(...categorySuggestions);
    });
    
    return relevantSuggestions;
}

/**
 * Main function to demonstrate fetching and using suggestions.
 */
async function main() {
    console.log("🔍 Fetching wellness suggestions from the API...");
    
    try {
        // Fetch all suggestions
        const suggestions = await fetchAllSuggestions();
        
        if (!suggestions) {
            console.log("❌ Failed to fetch suggestions. Make sure the API is running.");
            return;
        }
        
        // Print all suggestions
        printSuggestions(suggestions);
        
        // Example: Get suggestions for a specific category
        console.log("\n" + "=".repeat(50));
        console.log("🎯 Example: Stress Management Suggestions");
        console.log("=".repeat(50));
        const stressSuggestions = getSuggestionsByCategory(suggestions, "💪 Stress Management");
        stressSuggestions.forEach((suggestion, index) => {
            console.log(`${index + 1}. ${suggestion}`);
        });
        
        // Example: Get a random suggestion
        console.log("\n" + "=".repeat(50));
        console.log("🎲 Random Suggestion of the Day");
        console.log("=".repeat(50));
        const randomSuggestion = getRandomSuggestion(suggestions);
        console.log(`💡 ${randomSuggestion}`);
        
        // Example: Get suggestions based on emotion
        console.log("\n" + "=".repeat(50));
        console.log("😊 Suggestions for Joy");
        console.log("=".repeat(50));
        const joySuggestions = getSuggestionsByEmotion(suggestions, 'joy');
        joySuggestions.slice(0, 5).forEach((suggestion, index) => {
            console.log(`${index + 1}. ${suggestion}`);
        });
        
        // Example: Save suggestions to a JSON file
        console.log("\n" + "=".repeat(50));
        console.log("💾 Saving suggestions to file...");
        try {
            fs.writeFileSync('wellness_suggestions.json', JSON.stringify(suggestions, null, 2));
            console.log("✅ Suggestions saved to 'wellness_suggestions.json'");
        } catch (error) {
            console.log(`❌ Error saving suggestions: ${error.message}`);
        }
        
    } catch (error) {
        console.error(`❌ Error: ${error.message}`);
    }
}

// Run the main function if this script is executed directly
if (require.main === module) {
    main();
}

// Export functions for use in other modules
module.exports = {
    fetchAllSuggestions,
    printSuggestions,
    getSuggestionsByCategory,
    getRandomSuggestion,
    getSuggestionsByEmotion
}; 